### PR TITLE
Add video background blur

### PR DESCRIPTION
**Pull request by theunafraid**
TODO 

- [x]  Fix no data source error
- [x] Add comments

Improvements:
- [ ] Remove scaling by adding second MediaStream captured at lower resolution.

### CODE REVIEW

**Comment by theunafraid** [comment]
> Do we need to store and pass the video blur as media type or video type? Why?
> 
> Can't we store and pass this information separately? Maybe another flag?

Yes, will pass as option as we discussed offline

File: modules/RTC/JitsiLocalTrack.js [hunk 1]
```diff
@@ -190,7 +213,7 @@ export default class JitsiLocalTrack extends JitsiTrack {
      * issues.
      */
     _initNoDataFromSourceHandlers() {
-        if (this.isVideoTrack() && this.videoType === VideoType.CAMERA) {
+        if (this.isVideoTrack() && (this.videoType === VideoType.CAMERA)) {
```
Comment (hristoterezov) [inline modules/RTC/JitsiLocalTrack.js]: Do we need this change?

File: modules/RTC/JitsiLocalTrack.js [hunk 2]
```diff
@@ -301,6 +324,60 @@ export default class JitsiLocalTrack extends JitsiTrack {
         }
     }
 
+    /**
+     * Starts the video blur effect
```
Comment (hristoterezov) [inline modules/RTC/JitsiLocalTrack.js]: Add returns line

File: modules/RTC/JitsiLocalTrack.js [hunk 3]
```diff
@@ -301,6 +324,60 @@ export default class JitsiLocalTrack extends JitsiTrack {
...
+            cont => RTCUtils.attachMediaStream(cont, this.stream));
+        this._addStreamToConferenceAsUnmute();
+        this.effects.push('blur');
+        Statistics.sendAnalytics(TRACK_VIDEOBLUR, { value: 'enabled' });
+
+        return Promise.resolve();
+    }
+
+    /**
+     * Stops the video blur effect
```
Comment (hristoterezov) [inline modules/RTC/JitsiLocalTrack.js]: Add returns line

File: modules/RTC/JitsiStreamBlurEffect.js [hunk 4]
```diff
@@ -0,0 +1,178 @@
+import JitsiStreamEffect from './JitsiStreamEffect';
```
Comment (hristoterezov) [inline modules/RTC/JitsiStreamBlurEffect.js]: Add JSDoc for this file

File: modules/RTC/JitsiStreamBlurEffect.js [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+                        for (let j = 0; j < this._width; ++j) {
+
+                            x = Math.floor(j * this._scaleFactor);
+                            y = Math.floor(i * this._scaleFactor);
+
+                            segData.data[(i * this._width) + j] = value.data[(y * value.width) + x];
+                        }
+                    }
+
+                    // set
```
Comment (hristoterezov) [inline modules/RTC/JitsiStreamBlurEffect.js]: I don't really see why we have this comment. We should remove it or update it.

File: modules/RTC/JitsiStreamEffect.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
+/**
+ *
```
Comment (hristoterezov) [inline modules/RTC/JitsiStreamEffect.js]: Add JSDoc.

File: modules/RTC/RTCUtils.js [hunk 7]
```diff
@@ -1386,7 +1386,7 @@ class RTCUtils extends Listenable {
             const umDevices = options.devices || [ 'audio', 'video' ];
             const requestedCaptureDevices = umDevices.filter(device =>
                 device === 'audio'
-                || (device === 'video' && browser.supportsVideo()));
+                || ((device === 'video') && browser.supportsVideo()));
```
Comment (hristoterezov) [inline modules/RTC/RTCUtils.js]: Do we need this change?

File: modules/RTC/JitsiLocalTrack.js [hunk 8]
```diff
@@ -424,11 +507,16 @@ export default class JitsiLocalTrack extends JitsiTrack {
                         logger.warn(
                             `${this}: video type has changed after unmute!`,
                             this.videoType, streamInfo.videoType);
-                        this.videoType = streamInfo.videoType;
```
Comment (hristoterezov) [inline modules/RTC/JitsiLocalTrack.js]: Why did you remove this line?

File: modules/RTC/RTC.js [hunk 9]
```diff
@@ -102,7 +102,8 @@ function _newCreateLocalTracks(mediaStreamMetaData = []) {
             sourceType,
             stream,
             track,
-            videoType: videoType || null
+            videoType,
```
Comment (hristoterezov) [inline modules/RTC/RTC.js]: Why did you remove the `null`? Aren't we checking for `null` somewhere?

File: modules/RTC/JitsiLocalTrack.js [hunk 10]
```diff
@@ -366,6 +443,11 @@ export default class JitsiLocalTrack extends JitsiTrack {
                 this.track.enabled = !muted;
             }
         } else if (muted) {
+
```
Comment (hristoterezov) [inline modules/RTC/JitsiLocalTrack.js]: Unnecessary empty line.

File: modules/RTC/JitsiLocalTrack.js [hunk 11]
```diff
@@ -69,6 +74,19 @@ export default class JitsiLocalTrack extends JitsiTrack {
             mediaType,
             videoType);
 
+        this.effects = effects;
+
+        this._videoBlurEnabled = this.effects.indexOf('blur') !== -1
+            && mediaType === MediaType.VIDEO;
+
+        this._jitsiStreamBlurEffect = null;
```
Comment (hristoterezov) [inline modules/RTC/JitsiLocalTrack.js]: You can leave this set to `undefined`, isn't it? Isn't that going to be better? WDYT?

File: modules/RTC/RTCUtils.js (hunk 7 above)
Comment (theunafraid) [reply modules/RTC/RTCUtils.js]: No, sorry forgot to revert that change

File: modules/RTC/JitsiLocalTrack.js (hunk 8 above)
Comment (theunafraid) [reply modules/RTC/JitsiLocalTrack.js]: By mistake, thought it was my change before adding the effects option

File: modules/RTC/RTC.js (hunk 9 above)
Comment (theunafraid) [reply modules/RTC/RTC.js]: By mistake, corrected

File: modules/RTC/JitsiLocalTrack.js (hunk 1 above)
Comment (theunafraid) [reply modules/RTC/JitsiLocalTrack.js]: Not needed

File: modules/RTC/JitsiLocalTrack.js (hunk 11 above)
Comment (theunafraid) [reply modules/RTC/JitsiLocalTrack.js]: Agree, undefined is better.

File: modules/RTC/JitsiLocalTrack.js [hunk 12]
```diff
@@ -20,9 +20,11 @@ import RTCEvents from '../../service/RTC/RTCEvents';
 import VideoType from '../../service/RTC/VideoType';
 import {
     TRACK_UNMUTED,
-    createNoDataFromSourceEvent
+    createNoDataFromSourceEvent,
+    TRACK_VIDEOBLUR
 } from '../../service/statistics/AnalyticsEvents';
 import Statistics from '../statistics/statistics';
+import JitsiStreamBlurEffect from './JitsiStreamBlurEffect';
```
Comment (hristoterezov) [inline modules/RTC/JitsiLocalTrack.js]: Could you please leave an empty line? This way it will be consistent with jitsi-meet imports...

File: modules/RTC/JitsiLocalTrack.js [hunk 13]
```diff
@@ -174,6 +191,11 @@ export default class JitsiLocalTrack extends JitsiTrack {
         RTCUtils.addListener(RTCEvents.DEVICE_LIST_WILL_CHANGE, this._onDeviceListWillChange);
 
         this._initNoDataFromSourceHandlers();
+
+        if (this._videoBlurEnabled === true && this._jitsiStreamBlurEffect) {
```
Comment (hristoterezov) [inline modules/RTC/JitsiLocalTrack.js]: I'm asking out of curiosity - Why don't you include this logic into the `if` that initialises  `this. _jitsiStreamBlurEffect`?

File: service/statistics/AnalyticsEvents.js [hunk 14]
```diff
@@ -216,6 +216,14 @@ export const ICE_STATE_CHANGED = 'ice.state.changed';
  */
 export const TRACK_UNMUTED = 'track.unmuted';
 
+/**
+ * Indicates that a track has video blur enabled
+ *
+ * Properties:
+ *      value: 'enabled' or 'disabled'
+ */
+export const TRACK_VIDEOBLUR = 'track.videoblur';
```
Comment (hristoterezov) [inline service/statistics/AnalyticsEvents.js]: Why don't you create an event creator function?

File: modules/RTC/JitsiLocalTrack.js [hunk 15]
```diff
@@ -301,6 +323,64 @@ export default class JitsiLocalTrack extends JitsiTrack {
...
     }
 
+    /**
+     * Starts the video blur effect
+     *
+     * @returns {Promise}
+     */
+    startVideoBlurEffect() {
+        if (this.getType() !== MediaType.VIDEO) {
+            return Promise.reject();
```
Comment (hristoterezov) [inline modules/RTC/JitsiLocalTrack.js]: Add `new Error(...)` as an argument

File: modules/RTC/RTC.js [hunk 16]
```diff
@@ -76,7 +76,7 @@ function createLocalTracks(tracksInfo, options) {
  *     sourceType: The desktopsharing source type
  * }}
  */
-function _newCreateLocalTracks(mediaStreamMetaData = []) {
+function _newCreateLocalTracks(mediaStreamMetaData = [], options = {}) {
```
Comment (hristoterezov) [inline modules/RTC/RTC.js]: Add `options` to the JSDoc.

File: modules/RTC/JitsiStreamBlurEffect.js [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+     */
+    constructor(stream) {
+
+        super(stream);
+
+        let n = 1;
+
+        // Original video stream settings
+        const firstVideoTrack = this._stream.getVideoTracks().shift();
+        const { height, frameRate, width } = firstVideoTrack.getSettings();
```
Comment (hristoterezov) [inline modules/RTC/JitsiStreamBlurEffect.js]: What happens if the browser doesn't support `getSettings`?

File: modules/RTC/RTC.js [hunk 18]
```diff
@@ -102,7 +102,8 @@ function _newCreateLocalTracks(mediaStreamMetaData = []) {
             sourceType,
             stream,
             track,
-            videoType: videoType || null
+            videoType: videoType || null,
+            effects: options.effects
```
Comment (hristoterezov) [inline modules/RTC/RTC.js]: Isn't that going to add the effect property even on audio tracks? 

Can't we make `obtainAudioAndVideoPermissions` to include it in `tracksInfo` array? This way we wouldn't need this parameter and the effect property will be assigned only to the correct `JitsiLocalTrack`.

File: modules/RTC/JitsiLocalTrack.js [hunk 19]
```diff
@@ -777,6 +871,14 @@ export default class JitsiLocalTrack extends JitsiTrack {
         // we aren't receiving any data from the source. We want to notify
         // the users for error if the stream is muted or ended on it's
         // creation.
+
+        // For video blur enabled use the original video stream
+        if (this._videoBlurEnabled === true) {
```
Comment (hristoterezov) [inline modules/RTC/JitsiLocalTrack.js]: How about the following implementation:
```
   const stream =  this._videoBlurEnabled ? this._originalVideoStream : this.stream;

        return stream.getTracks().some(track =>
            (!('readyState' in track) || track.readyState === 'live')
                && (!('muted' in track) || track.muted !== true));
```

Comment (hristoterezov) [reply modules/RTC/JitsiLocalTrack.js]: Also in this case I guess you have to check if this._originalVideoStream exists...

File: modules/RTC/JitsiLocalTrack.js [hunk 20]
```diff
@@ -320,6 +342,64 @@ export default class JitsiLocalTrack extends JitsiTrack {
...
+    }
+
+    /**
+     * Stops the video blur effect
+     *
+     * @returns {Promise}
+     */
+    stopVideoBlurEffect() {
+        if (this.getType() !== MediaType.VIDEO) {
+            return Promise.reject();
```
Comment (hristoterezov) [inline modules/RTC/JitsiLocalTrack.js]: Add new Error(...) as an argument

File: modules/RTC/JitsiLocalTrack.js [hunk 21]
```diff
@@ -320,6 +342,64 @@ export default class JitsiLocalTrack extends JitsiTrack {
...
+            return Promise.resolve();
+        }
+
+        this._videoBlurEnabled = false;
+        if (this._jitsiStreamBlurEffect) {
+            this._jitsiStreamBlurEffect.stopEffect();
+        }
+
+        this._setStream(this._originalVideoStream);
+        this.containers.map(
```
Comment (hristoterezov) [inline modules/RTC/JitsiLocalTrack.js]: why `map` and not `forEach`?

File: modules/RTC/JitsiLocalTrack.js [hunk 22]
```diff
@@ -320,6 +342,64 @@ export default class JitsiLocalTrack extends JitsiTrack {
...
+            cont => RTCUtils.attachMediaStream(cont, this.stream));
+        this._addStreamToConferenceAsUnmute();
+        this.effects = this.effects.filter(flt => flt !== 'blur');
```
Comment (hristoterezov) [inline modules/RTC/JitsiLocalTrack.js]: why do we need this?

**Comment by hristoterezov** [comment]
@theunafraid It seems if you change the blur setting during video mute your video is not muted anymore and the video mute stops working.

**Comment by hristoterezov** [comment]
@theunafraid Another bug switching the state doesn't seem to work properly. I've joined with 2 participant - A and B. 

Bug#1:
Steps to reproduce:
1) A joins the call
2) B joins the call
3) A starts the video blur

Expected result:
B receives the blurred video.

Actual result:
B receives not blurred video.

Bug#2:
1) A joins the call
2) A enables the video blur
3) B joins the call and receives the blurred video
4) A disables the blurred video

Expected result:
B receives the original video.

Actual result:
B doesn't receive video.

**Comment by hristoterezov** [comment]
@theunafraid I've noticed that with your PR the size of lib-jitsi-meet bundle is 1.7 MB and before that it was ~600KB. This is huge increase. Any idea if we can fix this?

**Comment by theunafraid** [comment]
@hristoterezov I think this is because of tensorflow model...Not sure how to fix at the moment

**Comment by hristoterezov** [comment]
@theunafraid Increasing the size of the bundle 3 times for one feature is a very serious issue. We should work on fixing this somehow.

File: modules/RTC/JitsiLocalTrack.js (hunk 19 above)
Comment (theunafraid) [reply modules/RTC/JitsiLocalTrack.js]: Renamed _originalVideoStream to _originalStream for general use so that no check is needed

File: modules/RTC/RTC.js (hunk 18 above)
Comment (theunafraid) [reply modules/RTC/RTC.js]: I thought that maybe in the future we need an effect/processing of audio stream also. And i can give and example if we discuss offline. We can move it there for audio/video media but then i think i have to adapt JitsiLocalTrack also.

File: service/statistics/AnalyticsEvents.js (hunk 14 above)
Comment (theunafraid) [reply service/statistics/AnalyticsEvents.js]: I used Statistics.sendAnalytics as I saw it being used without a creator function - in RTCUtils.js

File: modules/RTC/JitsiStreamBlurEffect.js (hunk 17 above)
Comment (theunafraid) [reply modules/RTC/JitsiStreamBlurEffect.js]: I will use getContraints as alternative but saw in RTC.js _newCreateLocalTracks the following call :         const { deviceId, facingMode } = track.getSettings();

File: modules/RTC/RTC.js (hunk 18 above)
Comment (theunafraid) [reply modules/RTC/RTC.js]: Also the effects option is used from outside lib-jitsi-meet using JitsiMeetJS.js createLocalTracks otherwise it is empty array

**Comment by theunafraid** [comment]
@hristoterezov Fixed the 3 found bugs

File: modules/RTC/RTC.js (hunk 18 above)
Comment (theunafraid) [reply modules/RTC/RTC.js]: From JSDoc of JitsiMeetJS.js createLocalTracks 
* @param options Object with properties / settings specifying the tracks
* which should be created. should be created or some additional
* configurations about resolution for example.

File: modules/RTC/RTC.js (hunk 16 above)
Comment (theunafraid) [reply modules/RTC/RTC.js]: Also added JSDoc for JitsiMeetJS.js createLocalTracks

File: modules/RTC/JitsiLocalTrack.js (hunk 13 above)
Comment (theunafraid) [reply modules/RTC/JitsiLocalTrack.js]: Moved code to initialization if branch

File: modules/RTC/RTC.js (hunk 18 above)
Comment (theunafraid) [reply modules/RTC/RTC.js]: Moved effects to trackInfo

**Comment by saghul** [comment]
FTR, I'd like to show my opposition to the direction this is taking. I find the size increate to be unaccetable for just this one feature.

This could be implemented as a separate bundle which implements the blur effect itself. What lib-jitsi-meet would need is a way to register transformations / filters / whatever for the local track.

**Comment by hristoterezov** [comment]
@saghul The last versions of the PR are addressing the issue you are referring to. We have separate bundle build in jitsi-meet and the actual filter implementation is loaded dynamically when we need it and passed to lib-jitsi-meet trough the API.

Are you sure you are looking into the latest version of the code?

**Comment by saghul** [comment]
@hristoterezov That's awesome to hear! Looks like I hadn't clicked on the "refresh" banner thing 🤦‍♂  sorry for the noise!

File: modules/RTC/JitsiLocalTrack.js [hunk 23]
```diff
@@ -320,6 +335,121 @@ export default class JitsiLocalTrack extends JitsiTrack {
         }
     }
 
+    /**
+     * Sets the effect type that is used to modify the original stream
+     * @param {*} effectType
+     */
+    setEffect(effectType) {
+        this.EffectType = effectType;
```
[the remark this answers is not in the document]
Comment (theunafraid) [reply modules/RTC/JitsiLocalTrack.js]: Done, renamed this.EffectType to this._effectType

File: modules/RTC/JitsiLocalTrack.js [hunk 24]
```diff
@@ -320,6 +335,121 @@ export default class JitsiLocalTrack extends JitsiTrack {
...
+        this._originalStream = this.stream;
+    }
+
+    /**
+     * Starts the effect
+     *
+     * @returns {Promise}
+     */
+    startEffect() {
```
[the remark this answers is not in the document]
Comment (theunafraid) [reply modules/RTC/JitsiLocalTrack.js]: Removed setEffect and added a parameter to startEffect

File: modules/RTC/JitsiLocalTrack.js [hunk 25]
```diff
@@ -449,6 +607,14 @@ export default class JitsiLocalTrack extends JitsiTrack {
                     throw new JitsiTrackError(TRACK_NO_STREAM_FOUND);
                 }
 
+                this._originalStream = this.stream;
+
+                if (this.effectEnabled) {
+                    this._streamEffect = new this._effect(this._originalStream);
+                    this._setStream(this._streamEffect.getStreamWithEffect());
+                    this._streamEffect.startEffect();
+                }
```
[the remark this answers is not in the document]
Comment (theunafraid) [reply modules/RTC/JitsiLocalTrack.js]: Created new function _streamEffectEnable which is used in many places in this file.

File: modules/RTC/JitsiLocalTrack.js [hunk 26]
```diff
@@ -385,6 +536,12 @@ export default class JitsiLocalTrack extends JitsiTrack {
                 this.track.enabled = !muted;
             }
         } else if (muted) {
+
+            if (this.effectEnabled) {
```
[the remark this answers is not in the document]
Comment (theunafraid) [reply modules/RTC/JitsiLocalTrack.js]: Used new function _streamEffectEnable.

File: modules/RTC/RTCUtils.js [hunk 27]
```diff
@@ -1409,7 +1410,8 @@ class RTCUtils extends Listenable {
                 mediaStreamsMetaData.push({
                     stream: videoStream,
                     track: videoStream.getVideoTracks()[0],
-                    videoType: VideoType.CAMERA
+                    videoType: VideoType.CAMERA,
+                    effectTypes: options.effects
```
[the remark this answers is not in the document]
Comment (theunafraid) [reply modules/RTC/RTCUtils.js]: Renamed to just effects and used property shorthand

File: modules/RTC/RTC.js [hunk 28]
```diff
@@ -83,7 +85,8 @@ function _newCreateLocalTracks(mediaStreamMetaData = []) {
             sourceType,
             stream,
             track,
-            videoType
+            videoType,
+            effectTypes
```
[the remark this answers is not in the document]
Comment (theunafraid) [reply modules/RTC/RTC.js]: Renamed to effects

File: modules/RTC/JitsiLocalTrack.js [hunk 29]
```diff
@@ -320,6 +337,140 @@ export default class JitsiLocalTrack extends JitsiTrack {
...
+        return startEffectPromise;
+    }
+
+    /**
+     * Stops the effect
+     *
+     * @returns {Promise}
+     */
+    stopEffect() {
+        if (this.effectEnabled === false) {
```
[the remark this answers is not in the document]
Comment (theunafraid) [reply modules/RTC/JitsiLocalTrack.js]: Created a new function _switchStream which contains most of the common code and replaced startEffect and stopEffect with enableEffect function.

File: modules/RTC/JitsiLocalTrack.js [hunk 30]
```diff
@@ -320,6 +337,140 @@ export default class JitsiLocalTrack extends JitsiTrack {
...
+            const modifiedStream = this._streamEffect.getStreamWithEffect();
+
+            this._setStream(modifiedStream);
+
+            if (!this.isAudioTrack()) {
+                this.containers.forEach(
+                    cont => RTCUtils.attachMediaStream(cont, this.stream));
+            }
+
+            promises.push(this._addStreamToConferenceAsUnmute());
```
[the remark this answers is not in the document]
Comment (theunafraid) [reply modules/RTC/JitsiLocalTrack.js]: Reworked the code and moved it to new function _switchStream.

File: modules/RTC/JitsiLocalTrack.js [hunk 31]
```diff
@@ -320,6 +337,140 @@ export default class JitsiLocalTrack extends JitsiTrack {
...
+        });
+
+        const startEffectPromise = Promise.all(promises);
```
[the remark this answers is not in the document]
Comment (theunafraid) [reply modules/RTC/JitsiLocalTrack.js]: Code from this function is not in _switchStream.

File: modules/RTC/JitsiLocalTrack.js [hunk 32]
```diff
@@ -320,6 +337,140 @@ export default class JitsiLocalTrack extends JitsiTrack {
...
+
+        startEffectPromise.then(() => {
```
[the remark this answers is not in the document]
Comment (theunafraid) [reply modules/RTC/JitsiLocalTrack.js]: Code is now in _switchStream.

File: modules/RTC/JitsiLocalTrack.js [hunk 33]
```diff
@@ -320,6 +337,140 @@ export default class JitsiLocalTrack extends JitsiTrack {
         }
     }
 
+    /**
+     * Starts the effect and sets the effect type that is used to modify the original stream
+     * @param {*} effect
```
[the remark this answers is not in the document]
Comment (theunafraid) [reply modules/RTC/JitsiLocalTrack.js]: Added description for effect parameter

File: modules/RTC/JitsiLocalTrack.js [hunk 34]
```diff
@@ -69,6 +72,20 @@ export default class JitsiLocalTrack extends JitsiTrack {
             mediaType,
             videoType);
 
+        const effect = effects.find(e => e.isEnabled(this));
+
+        this.effectEnabled = false;
+        this.streamChangeInProgress = false;
```
[the remark this answers is not in the document]
Comment (theunafraid) [reply modules/RTC/JitsiLocalTrack.js]: Added _ to both

File: modules/RTC/JitsiLocalTrack.js [hunk 35]
```diff
@@ -42,6 +43,7 @@ export default class JitsiLocalTrack extends JitsiTrack {
      * JitsiRemoteTrack
      * @param trackInfo.mediaType the MediaType of the JitsiRemoteTrack
      * @param trackInfo.videoType the VideoType of the JitsiRemoteTrack
+     * @param trackInfo.effects the effect applied to the original stream
```
[the remark this answers is not in the document]
Comment (theunafraid) [reply modules/RTC/JitsiLocalTrack.js]: Updated the comment

File: modules/RTC/JitsiLocalTrack.js [hunk 36]
```diff
@@ -69,6 +72,20 @@ export default class JitsiLocalTrack extends JitsiTrack {
             mediaType,
             videoType);
 
+        const effect = effects.find(e => e.isEnabled(this));
+
+        this.effectEnabled = false;
```
[the remark this answers is not in the document]
Comment (theunafraid) [reply modules/RTC/JitsiLocalTrack.js]: Added _

File: modules/RTC/JitsiLocalTrack.js (hunk 10 above)
Comment (theunafraid) [reply modules/RTC/JitsiLocalTrack.js]: Removed empty line

File: modules/RTC/JitsiLocalTrack.js [hunk 37]
```diff
@@ -59,7 +61,8 @@ export default class JitsiLocalTrack extends JitsiTrack {
         sourceType,
         stream,
         track,
-        videoType
+        videoType,
+        effectTypes = []
```
[the remark this answers is not in the document]
Comment (theunafraid) [reply modules/RTC/JitsiLocalTrack.js]: Renamed to effects

File: modules/RTC/JitsiLocalTrack.js (hunk 24 above)
[the remark this answers is not in the document]
Comment (theunafraid) [reply modules/RTC/JitsiLocalTrack.js]: Replaced startEffect and stopEffect with enableEffect function

File: modules/RTC/JitsiLocalTrack.js [hunk 38]
```diff
@@ -320,6 +335,99 @@ export default class JitsiLocalTrack extends JitsiTrack {
...
+
+        return removeStreamPromise.then(() => {
+            this._setStream(newStream);
+
+            if (!this.isAudioTrack()) {
+                this.containers.forEach(
+                    cont => RTCUtils.attachMediaStream(cont, this.stream));
+            }
+
+            return this._addStreamToConferenceAsUnmute();
```
Comment (hristoterezov) [inline modules/RTC/JitsiLocalTrack.js]: `_removeLocalTrackAsMute` and `_addStreamToConferenceAsUnmute` are specifically designed for the video mute use case. Maybe we can create a new method that will replace the webrtc stream for a JitsiLocalTrack. WDYT? Probably we can do this even without Offer/Answer cycle.

File: modules/RTC/JitsiLocalTrack.js [hunk 39]
```diff
@@ -320,6 +335,99 @@ export default class JitsiLocalTrack extends JitsiTrack {
...
+    }
+
+    /**
+     * Switches the video stream associated with this track and updates the containers.
+     *
+     * @private
+     * @param {MediaStream} newStream - The MediaStream to replace
+     * @returns {Promise}
+     */
+    _switchStream(newStream: MediaStream) {
```
Comment (hristoterezov) [inline modules/RTC/JitsiLocalTrack.js]: Since we use `_switchStream` only in `enableEffect`, why don't we include it's code in `enableEffect`?

File: modules/RTC/JitsiLocalTrack.js [hunk 40]
```diff
@@ -320,6 +335,99 @@ export default class JitsiLocalTrack extends JitsiTrack {
...
 
+    /**
+     * Starts or stops the effect process and returns the corresponding stream.
+     *
+     * @private
+     * @param {Boolean} enable - Used to start or stop the effect processing
+     * @param {*} effect - Optional parameter represents effect class to create instance
+     * @returns {MediaStream}
+     */
+    _streamEffectEnable(enable: Boolean, effect) {
```
Comment (hristoterezov) [inline modules/RTC/JitsiLocalTrack.js]: Since the code is totally different for `enable===true` and `enable===false` I think it makes sense to have to methods. Maybe `_startStreamEffect` and `_stopStreamEffect`. WDYT?

File: modules/RTC/JitsiLocalTrack.js (hunk 39 above)
Comment (theunafraid) [reply modules/RTC/JitsiLocalTrack.js]: I was thinking that since it is a generic implementation maybe it can be used in the future

File: modules/RTC/JitsiLocalTrack.js [hunk 41]
```diff
@@ -449,6 +562,12 @@ export default class JitsiLocalTrack extends JitsiTrack {
                     throw new JitsiTrackError(TRACK_NO_STREAM_FOUND);
                 }
 
+                if (this._effectEnabled) {
+                    const modifiedStream = this._streamEffectEnable(true, this._effect);
+
+                    this._setStream(modifiedStream);
```
[the remark this answers is not in the document]
Comment (theunafraid) [reply modules/RTC/JitsiLocalTrack.js]: I agree, but i thought that we need the code to be more readable

File: modules/RTC/JitsiLocalTrack.js (hunk 38 above)
Comment (theunafraid) [reply modules/RTC/JitsiLocalTrack.js]: I could also use removeTrack and addTrack from JitsiConference to get the same result

File: modules/RTC/JitsiLocalTrack.js (hunk 40 above)
Comment (theunafraid) [reply modules/RTC/JitsiLocalTrack.js]: Create the two methods _startStreamEffect and _stopStreamEffect

File: modules/RTC/JitsiLocalTrack.js [hunk 42]
```diff
@@ -536,6 +655,12 @@ export default class JitsiLocalTrack extends JitsiTrack {
      * @returns {Promise}
      */
     dispose() {
+        if (this.effectEnabled) {
```
[the remark this answers is not in the document]
Comment (theunafraid) [reply modules/RTC/JitsiLocalTrack.js]: Fixed

File: modules/RTC/RTC.js [hunk 43]
```diff
@@ -53,7 +53,8 @@ function createLocalTracks(tracksInfo, options) {
             ...trackInfo,
             deviceId,
             facingMode: options.facingMode,
-            rtcId: rtcTrackIdCounter
+            rtcId: rtcTrackIdCounter,
+            effect: options.effects
```
[the remark this answers is not in the document]
Comment (theunafraid) [reply modules/RTC/RTC.js]: Fixed

File: modules/RTC/JitsiLocalTrack.js (hunk 38 above)
Comment (hristoterezov) [reply modules/RTC/JitsiLocalTrack.js]: removeTrack and addTrack is fine for now. As further optimisation we can create JingleSession method that replaces a stream without doing Offer Answer. But let's address this trough another PR when we have more time.

File: modules/RTC/JitsiLocalTrack.js [hunk 44]
```diff
@@ -306,6 +319,103 @@ export default class JitsiLocalTrack extends JitsiTrack {
...
+            return Promise.reject(new Error('Stream change already in progress!'));
+        }
+
+        this._streamChangeInProgress = true;
+
+        if (!this.conference) {
+            return Promise.resolve();
+        }
+
+        const conference = this.conference;
```
[the remark this answers is not in the document]
Comment (theunafraid) [reply modules/RTC/JitsiLocalTrack.js]: But this way the assignment will not be executed because of the earlier return

[the remark this answers is not in the document]
Comment (hristoterezov) [reply modules/RTC/JitsiLocalTrack.js]: Yes but we'll have 2 times the operation to search for the property... I don't think there's really a big performance tradeoff. I just like the solution for the early assignment of the `conference` more. If you don't like it you can keep the code as it is.
